storage/dataflux/fast_list.go

```diff
@@ -73,18 +86,68 @@ type Lister struct {
 type CloseFunc func()
 
 // NewLister creates a new dataflux Lister to list objects in the give bucket.
-func NewLister(opts *Input) (*Lister, CloseFunc) {
-
-	lister := &Lister{}
+func NewLister(c *storage.Client, in *ListerInput) (*Lister, CloseFunc) {
+	bucket := c.Bucket(in.BucketName)
+	lister := &Lister{
+		method:               open,
+		pageToken:            "",
+		bucket:               bucket,
+		batchSize:            in.BatchSize,
+		query:                in.Query,
+		skipDirectoryObjects: in.SkipDirectoryObjects,
+	}
 	return lister, func() { lister.Close() }
 }
 
 // NextBatch runs worksteal algorithm and sequential listing in parallel to quickly
-// return a list of objects in the bucket. For buckets with smaller dataset,
-// sequential listing is expected to be faster. For buckets with larger dataset,
+// return a list of objects in the bucket. For smaller dataset,
+// sequential listing is expected to be faster. For larger dataset,
 // worksteal listing is expected to be faster.
 func (c *Lister) NextBatch(ctx context.Context) ([]*storage.ObjectAttrs, error) {
-	return nil, nil
+	// countError tracks the number of failed listing methods.
+	countError := 0
+	var results []*storage.ObjectAttrs
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	g, childCtx := errgroup.WithContext(ctx)
```

Comment:
I'm not sure errgroup is what you want here. IIUC this means if one listing method fails, then the other will be cancelled as well. In addition, you probably don't want to wait on results from both methods, but rather take the first to succeed. I think you probably need to manually spin up go routines for each method and then have a channel to send results on, then wait on that channel.

I think this is fine to leave in for the MVP since you are only supporting one listing method so far, but you should add TODOs to fix this.

Reply:
>>  you probably don't want to wait on results from both methods, but rather take the first to succeed. 

Errgroup takes care of running both methods in parallel. I am cancelling context which stops the running method as soon as one of them is complete(fail/succeed). So we don't wait for both the methods to complete.

>>  if one listing method fails, then the other will be cancelled as well.

It is true, as soon as one of the method fails, the other method will fail. 
If you see sequential listing it will only fail due to some error while iterating objects. Worksteal listing uses similar iterator so there is a good chance that worksteal will fail too. 
If worksteal fails, I think it would make sense to show the error instead of silently falling back to sequential listing as user would not be able to take advantage of fast-listing then.

Reply:
Ahh, thanks for the explanation. That makes sense. Maybe just note the thing about not falling back to the other method in a comment.